pkg/controller/scheduledscaling/scheduled_scaling.go

```diff
@@ -183,30 +191,220 @@ func (c *Controller) runOnce(ctx context.Context) error {
+	if err != nil {
+		return fmt.Errorf("failed to adjust scaling: %w", err)
+	}
+
+	return nil
+}
+
+// activeScheduledScaling returns a map of the current active schedules and the
+// max value per schedule.
+func (c *Controller) activeScheduledScaling(schedules []*v1.ScalingSchedule, clusterschedules []*v1.ClusterScalingSchedule) map[string]int64 {
+	currentActiveSchedules := make(map[string]int64)
+
+	// ScalingSchedules
+	for _, schedule := range schedules {
+		activeSchedules, err := c.activeSchedules(schedule.Spec)
+		if err != nil {
+			log.Errorf("Failed to check for active schedules in ScalingSchedule %s/%s: %v", schedule.Namespace, schedule.Name, err)
+			continue
+		}
+
+		if len(activeSchedules) == 0 {
+			continue
+		}
+
+		maxValue := int64(0)
+		for _, activeSchedule := range activeSchedules {
+			if activeSchedule.Value > maxValue {
+				maxValue = activeSchedule.Value
+			}
+		}
+		currentActiveSchedules[schedule.Namespace+"/"+schedule.Name] = maxValue
+	}
+
+	for _, schedule := range clusterschedules {
+		activeSchedules, err := c.activeSchedules(schedule.Spec)
+		if err != nil {
+			log.Errorf("Failed to check for active schedules in ClusterScalingSchedule %s: %v", schedule.Name, err)
+			continue
+		}
+
+		if len(activeSchedules) == 0 {
+			continue
+		}
+
+		maxValue := int64(0)
+		for _, activeSchedule := range activeSchedules {
+			if activeSchedule.Value > maxValue {
+				maxValue = activeSchedule.Value
+			}
+		}
+		currentActiveSchedules[schedule.Name] = maxValue
+	}
+
+	return currentActiveSchedules
+}
+
+// adjustHPAScaling adjusts the scaling for a single HPA based on the active
+// scaling schedules. An adjustment is made if the current HPA scale is below
+// the desired and the change is within the HPA tolerance.
+func (c *Controller) adjustHPAScaling(ctx context.Context, hpa *autoscalingv2.HorizontalPodAutoscaler, activeSchedules map[string]int64) error {
+	current := int64(hpa.Status.CurrentReplicas)
+	if current == 0 {
+		return nil
+	}
+
+	var highestExpected int64
+	var highestObject autoscalingv2.CrossVersionObjectReference
+	for _, metric := range hpa.Spec.Metrics {
+		if metric.Type != autoscalingv2.ObjectMetricSourceType {
+			continue
+		}
+
+		switch metric.Object.DescribedObject.Kind {
+		case "ClusterScalingSchedule", "ScalingSchedule":
+		default:
+			continue
+		}
+
+		scheduleName := metric.Object.DescribedObject.Name
+
+		target := int64(metric.Object.Target.AverageValue.MilliValue() / 1000)
+		if target == 0 {
+			continue
+		}
+
+		var value int64
+		switch metric.Object.DescribedObject.Kind {
+		case "ScalingSchedule":
+			value = activeSchedules[hpa.Namespace+"/"+scheduleName]
+		case "ClusterScalingSchedule":
+			value = activeSchedules[scheduleName]
+		}
+
+		expected := int64(math.Ceil(float64(value) / float64(target)))
+		if expected > highestExpected {
+			highestExpected = expected
+			highestObject = metric.Object.DescribedObject
+		}
+	}
```

Comment:
Let's break this out into a function like `getHighestExpected` or something similar. Seems to be an independent piece of code we can refactor easily.

Reply:
Moved to a separate function.